File: ui/src/lib/inviteHtml.ts
/**
 * inviteHtml — email-safe HTML render helpers for the invitation
 * blocks that "Respond with meeting" and "Insert Talk link" drop
 * into an outgoing message body (#195).
 *
 * Constraints we design around:
 *   • Inline styles only.  Many mail clients strip <style>
 *     blocks, classes carry no meaning across clients.
 *   • System font stack — no @import, no remote @font-face.
 *   • Width capped at 560 px so the card doesn't blow out a
 *     mobile inbox view.
 *   • Detail-row glyphs are emoji (📅 🕐 📍 📝) because every
 *     client renders them; SVG / icon fonts inside email are
 *     unreliable across mail clients.
 *   • Banner image referenced via the project's GitHub raw URL
 *     so the recipient's mail client fetches a real PNG of the
 *     Unkai logo when they permit remote content.  Falls back
 *     gracefully to alt-text if blocked.
 *
 * The visual language across both cards is identical: same
 * gradient header, same surface card, same CTA button shape.
 * That keeps a thread carrying a Talk room *and* a meeting card
 * (the common case) reading as one consistent invitation,
 * not two unrelated stickers.
 */

/** Compatibility export — the brand logo lived here as a data
 *  URI for one cycle, before we discovered that many mail
 *  clients and corporate filters strip `<img src="data:…">`
 *  (and any remote URL hits remote-image blocking on first
 *  read). Both paths produced a broken-icon artifact for the
 *  recipient, so the chrome dropped the `<img>` entirely in
 *  favour of a typography-only wordmark.
 *
 *  Kept as an empty string so any external import resolves to
 *  something falsy without breaking the bundle; new code
 *  shouldn't reference it. Remove on next cleanup pass once we
 *  confirm nothing else imports it. */
export const PUBLIC_UNKAI_LOGO_URL = ''
/** Minimal HTML escape for the few values we splice into the
 *  template. Same characters DOMPurify-tier sanitisers cover —
 *  we don't need full sanitisation here because the inputs are
 *  app-side strings (event summary, room name, attendee list)
 *  rather than untrusted remote content. */
function esc(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;')
}

/** Format an ISO / Date pair as a single human line:
 *  "Tuesday, May 6 · 14:00 – 15:00".  Locale-aware via Intl,
 *  weekday-first because that's how every mature calendar
 *  invitation reads. */
function formatRange(start: Date, end: Date): string {
  const sameDay =
    start.getFullYear() === end.getFullYear() &&
    start.getMonth() === end.getMonth() &&
    start.getDate() === end.getDate()
  const dateFmt = new Intl.DateTimeFormat(undefined, {
    weekday: 'long',
    month: 'long',
    day: 'numeric',
  })
  const timeFmt = new Intl.DateTimeFormat(undefined, {
    hour: '2-digit',
    minute: '2-digit',
  })
  const datePart = dateFmt.format(start)
  if (sameDay) {
    return `${datePart} · ${timeFmt.format(start)} – ${timeFmt.format(end)}`
  }
  return `${datePart} ${timeFmt.format(start)} – ${dateFmt.format(end)} ${timeFmt.format(end)}`
}

// ── Shared design tokens (inlined into every style attribute) ──
//
// Tokens kept as string constants so a future tweak to the
// invite palette stays a single-file edit.  Values picked to
// read cleanly on both light and dark mail-client backgrounds —
// the card sits on its own white surface regardless.

const T = {
  card: 'background:#ffffff;border:1px solid #e2e8f0;border-radius:14px;box-shadow:0 1px 2px rgba(15,23,42,0.04),0 8px 24px rgba(15,23,42,0.06);overflow:hidden;',
  headerBg:
    'background:linear-gradient(135deg,#3b82f6 0%,#6366f1 100%);padding:20px 24px;color:#ffffff;',
  // Typography-only header: a wordmark inside a soft pill.
  // Many mail clients strip `<img src="data:…">` URIs for
  // security, and the previous remote-URL approach hit image-
  // blocking by default — both left users staring at a broken-
  // image icon. The pill + wordmark renders identically in every
  // client because it's pure inline-styled HTML.
  headerPill:
    'display:inline-block;padding:7px 14px;border-radius:999px;background:rgba(255,255,255,0.18);',
  headerWordmark:
    'font-size:13px;font-weight:700;letter-spacing:0.12em;color:#ffffff;text-transform:uppercase;',
  body: 'padding:24px;font-family:-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;color:#0f172a;',
  title:
    'margin:0 0 4px 0;font-size:20px;line-height:1.3;font-weight:600;color:#0f172a;',
  subtitle:
    'margin:0;font-size:14px;line-height:1.5;color:#475569;',
  detailLabel:
    'font-size:11px;font-weight:600;letter-spacing:0.06em;text-transform:uppercase;color:#64748b;',
  detailValue:
    'font-size:14px;line-height:1.5;color:#0f172a;margin:2px 0 0 0;',
  divider: 'height:1px;background:#e2e8f0;margin:20px 0;border:0;',
  ctaRow: 'margin-top:24px;',
  ctaButton:
    'display:inline-block;background:#3b82f6;color:#ffffff;text-decoration:none;font-weight:600;font-size:14px;padding:11px 22px;border-radius:10px;letter-spacing:0.01em;',
  footer:
    'margin:20px 0 0 0;padding:14px 16px;background:#f8fafc;border-radius:10px;font-size:12px;line-height:1.5;color:#475569;',
  footerStrong: 'color:#0f172a;font-weight:600;',
} as const

/** Wrap the card body inside the shared chrome (outer wrapper +
 *  branded header). The header is a typography-only wordmark in
 *  a soft pill — no image, because many mail clients and
 *  corporate filters strip `<img src="data:…">` for security,
 *  and remote URLs hit image-blocking by default, so any image-
 *  based brand element ends up as a broken-icon artifact for the
 *  recipient. The wordmark renders identically in every client.
 *
 *  The outer `data-unkai-block` attribute is the marker the
 *  RichTextEditor's `UnkaiBlock` extension parses into an atom
 *  node — that's how the styled card survives Tiptap's schema
 *  unwrapping when the HTML is loaded into the editor. */
function chrome(bodyHtml: string, kind: string): string {
  return `
<div data-unkai-block="${kind}" style="max-width:560px;margin:16px 0;font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Helvetica,Arial,sans-serif;">
  <div style="${T.card}">
    <div style="${T.headerBg}">
      <span style="${T.headerPill}"><span style="${T.headerWordmark}">Unkai Mail</span></span>
    </div>
    <div style="${T.body}">
      ${bodyHtml}
    </div>
  </div>
</div>
`.trim()
}

/** One detail row: emoji + label + value.  Stacked vertically
 *  so the wrap-around case (long location, long Talk URL) reads
 *  cleanly instead of forcing horizontal scroll. */
function detailRow(emoji: string, label: string, value: string): string {
  return `
<div style="display:flex;gap:12px;align-items:flex-start;margin:14px 0;">
  <div style="font-size:18px;line-height:1.4;flex-shrink:0;width:24px;text-align:center;">${emoji}</div>
  <div style="flex:1;min-width:0;">
    <div style="${T.detailLabel}">${esc(label)}</div>
    <div style="${T.detailValue}">${value}</div>
  </div>
</div>
`.trim()
}

/** Footer microcopy block — the "You'll be invited via
 *  Nextcloud" line lives here so it reads as a system-level
 *  note rather than competing with the meeting title. */
function nextcloudFooter(message: string): string {
  return `
<div style="${T.footer}">
  <span style="${T.footerStrong}">📨 ${esc(message)}</span><br />
  <span>Accept the invitation in your mail client or directly in Nextcloud — your calendar updates either way.</span>
</div>
`.trim()
}

// ── Public renderers ────────────────────────────────────────────

export interface TalkInvite {
  /** Display name of the Talk room. */
  name: string
  /** Full join URL the recipient will click. */
  url: string
}

/** Render-time options for the invite cards.  Currently empty —
 *  the chrome no longer carries a logo image, so there's nothing
 *  to override.  Kept as an exported type so existing call sites
 *  that pass `{}` keep type-checking and we have a place to add
 *  knobs later (theme override, etc.) without churning every
 *  signature. */
export type InviteRenderOptions = Record<string, never>

/** Render a Talk-meeting invitation card.  Used by Compose's
 *  "Insert Talk link" flow when the user attaches a freshly-
 *  created Talk room to an outgoing message. */
export function talkInviteHtml(invite: TalkInvite, _opts: InviteRenderOptions = {}): string {
  const inner = `
<h1 style="${T.title}">You're invited to a Talk meeting</h1>
<p style="${T.subtitle}">Click the button below to join the conversation in Nextcloud Talk — works in any modern browser, no install.</p>

<hr style="${T.divider}" />

${detailRow(
  '💬',
  'Talk room',
  `<strong style="font-weight:600;">${esc(invite.name)}</strong>`,
)}
${detailRow(
  '🔗',
  'Join link',
  `<a href="${esc(invite.url)}" style="color:#3b82f6;text-decoration:none;word-break:break-all;">${esc(invite.url)}</a>`,
)}

<div style="${T.ctaRow}">
  <a href="${esc(invite.url)}" style="${T.ctaButton}">Join Talk meeting →</a>
</div>

${nextcloudFooter("You'll also be added as a participant in Nextcloud Talk.")}
`.trim()
  return chrome(inner, 'talk-invite')
}

export interface MeetingInvite {
  /** Event title / summary line. */
  summary: string
  /** Event start (Date or ISO string the constructor accepts). */
  start: Date | string
  /** Event end (Date or ISO string). */
  end: Date | string
  /** Optional location text (free-form — a room name, address, or
   *  Talk URL the user typed into the Location field). */
  location?: string | null
  /** Optional notes / description.  Plain text; we wrap-preserve
   *  newlines so a multi-line agenda survives the render. */
  description?: string | null
  /** Optional Talk room URL.  Rendered as its own row + a second
   *  CTA button so the recipient can join the conversation
   *  ahead of the meeting if they want. */
  talkUrl?: string | null
}

/** Render a calendar-meeting invitation card.  Used after the
 *  user saves an event from the "Respond with meeting" flow —
 *  Compose opens with this block pre-filled in the body. */
export function meetingInviteHtml(invite: MeetingInvite, _opts: InviteRenderOptions = {}): string {
  const start = invite.start instanceof Date ? invite.start : new Date(invite.start)
  const end = invite.end instanceof Date ? invite.end : new Date(invite.end)

  const detailRows: string[] = [
    detailRow(
      '📅',
      'When',
      `<strong style="font-weight:600;">${esc(formatRange(start, end))}</strong>`,
    ),
  ]
  if (invite.location && invite.location.trim()) {
    detailRows.push(detailRow('📍', 'Where', esc(invite.location.trim())))
  }
  if (invite.talkUrl && invite.talkUrl.trim()) {
    detailRows.push(
      detailRow(
        '💬',
        'Talk room',
        `<a href="${esc(invite.talkUrl)}" style="color:#3b82f6;text-decoration:none;word-break:break-all;">${esc(invite.talkUrl)}</a>`,
      ),
    )
  }
  if (invite.description && invite.description.trim()) {
    // Preserve plain-text newlines as <br> so an agenda with
    // line breaks reads correctly.  The text content goes
    // through `esc` first so any `<` `>` `&` in the user's
    // notes don't accidentally break the surrounding HTML.
    const notesHtml = esc(invite.description.trim()).replace(/\r?\n/g, '<br />')
    detailRows.push(detailRow('📝', 'Notes', notesHtml))
  }

  const ctaHref = invite.talkUrl && invite.talkUrl.trim() ? invite.talkUrl : null
  const ctaBlock = ctaHref
    ? `
<div style="${T.ctaRow}">
  <a href="${esc(ctaHref)}" style="${T.ctaButton}">Join Talk meeting →</a>
</div>
`.trim()
    : ''

  const inner = `
<h1 style="${T.title}">${esc(invite.summary)}</h1>
<p style="${T.subtitle}">A calendar invitation has been created in Nextcloud and shared with everyone on this thread.</p>

<hr style="${T.divider}" />

${detailRows.join('\n')}

${ctaBlock}

${nextcloudFooter("You'll be invited via Nextcloud — accepting in your mail client adds the event to your calendar.")}
`.trim()
  return chrome(inner, 'meeting-invite')
}

// ── Quoted-history wrapper (#195 follow-up) ─────────────────────

/** Wrap a reply's quoted history in a styled container.  Visually
 *  pushes the previous thread back so the user's fresh reply
 *  reads first: muted background, smaller type, soft left border,
 *  generous padding.  Same subdued grey palette across light and
 *  dark themes — designed to read as "old context" rather than
 *  competing with the new content.
 *
 *  The outer `data-unkai-block` attribute lets the editor's
 *  `UnkaiBlock` extension capture this as an atom node so the
 *  styled wrapper survives Tiptap's schema (which would otherwise
 *  unwrap the `<div>` and strip the inline styles).
 *
 *  The inner content is also wrapped in a `<blockquote
 *  type="cite">` (#277).  Standard mail clients use blockquote
 *  as the canonical "this is quoted from a previous mail"
 *  signal — Apple Mail, Thunderbird, Gmail, Outlook each render
 *  blockquoted content with their own indent / collapse
 *  affordance ("Show original", a vertical bar, etc.).  The
 *  `type="cite"` attribute is the long-standing convention for
 *  email-quoted blockquotes (vs. literary citations); it does
 *  nothing visual but a couple of clients use it as a stronger
 *  signal that the block is a mail quote.  We zero out the
 *  blockquote's default browser indent so it doesn't double
 *  up with the styled div's own padding. */
export function quotedHistoryHtml(args: {
  fromHeader: string
  whenText: string
  bodyHtml: string
}): string {
  const { fromHeader, whenText, bodyHtml } = args
  const escAttr = (s: string) =>
    s
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&#39;')
  const meta = `On ${escAttr(whenText)}, ${escAttr(fromHeader)} wrote:`
  return `
<div data-unkai-block="quoted-history" style="margin:24px 0 0 0;padding:14px 18px;background:#f1f5f9;border-left:3px solid #cbd5e1;border-radius:8px;color:#64748b;font-size:13px;line-height:1.55;font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Helvetica,Arial,sans-serif;">
  <div style="font-size:11px;font-weight:600;letter-spacing:0.06em;text-transform:uppercase;color:#94a3b8;margin-bottom:8px;">Previous conversation</div>
  <div style="font-size:12px;color:#475569;margin-bottom:10px;">${meta}</div>
  <blockquote type="cite" style="margin:0;padding:0;border:0;color:#64748b;">${bodyHtml}</blockquote>
</div>
`.trim()
}

// ── Forwarded-mail wrapper (#319) ───────────────────────────────

/** Wrap a forwarded message's body so the original formatting
 *  is preserved verbatim and the header matches the long-standing
 *  forwarded-mail convention every receiving mail client knows.
 *
 *  The header is the de-facto-standard plain-text delimiter
 *  (`---------- Forwarded message ----------`) followed by
 *  bold-labelled `From:` / `Date:` / `Subject:` / `To:` lines.
 *  This is the same shape virtually every other mail client emits,
 *  so the recipient's client renders the forward the way they
 *  expect rather than seeing a bespoke chrome from us.
 *
 *  The `data-unkai-block` wrapper carries no visual styling — it
 *  exists purely so the editor's `UnkaiBlock` extension captures
 *  the whole chunk as an atom node. Without that, Tiptap's
 *  StarterKit schema unwraps the wrapper, strips the original
 *  message's inline styles, and can duplicate `<img>` nodes while
 *  reconciling complex `<table>` / `<picture>` markup against its
 *  block schema (the original visual symptom on #319). Inside the
 *  atom Tiptap doesn't parse the interior, so the email's table
 *  layout, inline styles, and `<img>` elements survive untouched.
 *  The chunk is non-editable; users add their note above it and
 *  remove it as a unit with one Backspace if they don't want it.
 *
 *  `toHeader` is optional because the field is occasionally
 *  missing on cached payloads — when absent we just omit the line
 *  rather than print an empty `To:`. */
export function forwardedMailHtml(args: {
  fromHeader: string
  whenText: string
  subjectText: string
  toHeader?: string
  bodyHtml: string
}): string {
  const { fromHeader, whenText, subjectText, toHeader, bodyHtml } = args
  const esc = (s: string) =>
    s
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
  const lines = [
    `<div><b>From:</b> ${esc(fromHeader)}</div>`,
    `<div><b>Date:</b> ${esc(whenText)}</div>`,
    `<div><b>Subject:</b> ${esc(subjectText)}</div>`,
  ]
  if (toHeader && toHeader.trim()) {
    lines.push(`<div><b>To:</b> ${esc(toHeader)}</div>`)
  }
  return `
<div data-unkai-block="forwarded-mail">
  <p>---------- Forwarded message ----------</p>
  ${lines.join('\n  ')}
  <p></p>
  ${bodyHtml}
</div>
`.trim()
}

File: crates/unkai-mcp/src/invite_html.rs
//! Rust port of the email-safe invite-card builders in
//! `ui/src/lib/inviteHtml.ts` (#441).
//!
//! `create_meeting_invite` produces the same invite-card draft the
//! in-app "Respond with meeting" / "Insert Talk link" flows do, so
//! this module mirrors the TypeScript builders **template for
//! template** — same wrapper structure, same inline-style tokens,
//! same `data-unkai-block` markers (which the editor's `UnkaiBlock`
//! extension needs to keep the card intact as an atom node).
//!
//! ## Keeping the two implementations in sync
//!
//! There is deliberately no shared source of truth: the TS side
//! renders inside the webview, this side inside the Rust process,
//! and a build-time bridge between them would cost more than the
//! duplication.  Instead the rule is: **any change to
//! `inviteHtml.ts`'s tokens or templates must be applied here too**
//! (and vice versa).  The unit tests below lock the exact byte
//! shape of both cards so an unsynced edit fails loudly.
//!
//! One knowing divergence: the TS side formats the date line with
//! `Intl.DateTimeFormat(undefined, …)` — the OS locale, including
//! its 12/24-hour convention.  Rust has no `Intl`; this port pins
//! English month/weekday names and 24-hour times ("Tuesday,
//! May 6 · 14:00 – 15:00").  The shape of the line is identical,
//! only the localisation differs.
//!
//! The email-rendering constraints (inline styles only, system
//! font stack, emoji glyphs, no images in the chrome) are
//! documented in `CLAUDE.md` — don't relax them here.

use chrono::{DateTime, Datelike, Local, TimeZone, Utc};

/// Minimal HTML escape for the values spliced into the templates —
/// same five characters the TS `esc` helper covers.
fn esc(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&#39;")
}

// ── Shared design tokens (inlined into every style attribute) ──
//
// Byte-identical to the `T` constant in `inviteHtml.ts`.

const T_CARD: &str = "background:#ffffff;border:1px solid #e2e8f0;border-radius:14px;box-shadow:0 1px 2px rgba(15,23,42,0.04),0 8px 24px rgba(15,23,42,0.06);overflow:hidden;";
const T_HEADER_BG: &str =
    "background:linear-gradient(135deg,#3b82f6 0%,#6366f1 100%);padding:20px 24px;color:#ffffff;";
const T_HEADER_PILL: &str =
    "display:inline-block;padding:7px 14px;border-radius:999px;background:rgba(255,255,255,0.18);";
const T_HEADER_WORDMARK: &str =
    "font-size:13px;font-weight:700;letter-spacing:0.12em;color:#ffffff;text-transform:uppercase;";
const T_BODY: &str = "padding:24px;font-family:-apple-system,BlinkMacSystemFont,\"Segoe UI\",Roboto,Helvetica,Arial,sans-serif;color:#0f172a;";
const T_TITLE: &str =
    "margin:0 0 4px 0;font-size:20px;line-height:1.3;font-weight:600;color:#0f172a;";
const T_SUBTITLE: &str = "margin:0;font-size:14px;line-height:1.5;color:#475569;";
const T_DETAIL_LABEL: &str =
    "font-size:11px;font-weight:600;letter-spacing:0.06em;text-transform:uppercase;color:#64748b;";
const T_DETAIL_VALUE: &str = "font-size:14px;line-height:1.5;color:#0f172a;margin:2px 0 0 0;";
const T_DIVIDER: &str = "height:1px;background:#e2e8f0;margin:20px 0;border:0;";
const T_CTA_ROW: &str = "margin-top:24px;";
const T_CTA_BUTTON: &str = "display:inline-block;background:#3b82f6;color:#ffffff;text-decoration:none;font-weight:600;font-size:14px;padding:11px 22px;border-radius:10px;letter-spacing:0.01em;";
const T_FOOTER: &str = "margin:20px 0 0 0;padding:14px 16px;background:#f8fafc;border-radius:10px;font-size:12px;line-height:1.5;color:#475569;";
const T_FOOTER_STRONG: &str = "color:#0f172a;font-weight:600;";

/// Format a start/end pair as the single human line the "When" row
/// carries: `Tuesday, May 6 · 14:00 – 15:00` (same-day) or
/// `Tuesday, May 6 14:00 – Wednesday, May 7 15:00` (cross-day).
/// English names + 24-hour clock — see the module comment for why
/// this pins a locale where the TS side follows the OS.
pub(crate) fn format_range<Tz: TimeZone>(start: &DateTime<Tz>, end: &DateTime<Tz>) -> String
where
    Tz::Offset: std::fmt::Display,
{
    let date = |d: &DateTime<Tz>| d.format("%A, %B %-d").to_string();
    let time = |d: &DateTime<Tz>| d.format("%H:%M").to_string();
    let same_day =
        start.year() == end.year() && start.month() == end.month() && start.day() == end.day();
    if same_day {
        format!("{} · {} – {}", date(start), time(start), time(end))
    } else {
        format!(
            "{} {} – {} {}",
            date(start),
            time(start),
            date(end),
            time(end)
        )
    }
}

/// Wrap the card body inside the shared chrome (outer wrapper +
/// typography-only wordmark header).  The `data-unkai-block`
/// attribute is what the editor's `UnkaiBlock` extension keys on.
fn chrome(body_html: &str, kind: &str) -> String {
    format!(
        "<div data-unkai-block=\"{kind}\" style=\"max-width:560px;margin:16px 0;font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Helvetica,Arial,sans-serif;\">\n  <div style=\"{T_CARD}\">\n    <div style=\"{T_HEADER_BG}\">\n      <span style=\"{T_HEADER_PILL}\"><span style=\"{T_HEADER_WORDMARK}\">Unkai Mail</span></span>\n    </div>\n    <div style=\"{T_BODY}\">\n      {body_html}\n    </div>\n  </div>\n</div>"
    )
}

/// One detail row: emoji + label + value, stacked vertically.
fn detail_row(emoji: &str, label: &str, value: &str) -> String {
    format!(
        "<div style=\"display:flex;gap:12px;align-items:flex-start;margin:14px 0;\">\n  <div style=\"font-size:18px;line-height:1.4;flex-shrink:0;width:24px;text-align:center;\">{emoji}</div>\n  <div style=\"flex:1;min-width:0;\">\n    <div style=\"{T_DETAIL_LABEL}\">{}</div>\n    <div style=\"{T_DETAIL_VALUE}\">{value}</div>\n  </div>\n</div>",
        esc(label)
    )
}

/// Footer microcopy block — the "You'll be invited via Nextcloud"
/// line lives here so it reads as a system-level note.
fn nextcloud_footer(message: &str) -> String {
    format!(
        "<div style=\"{T_FOOTER}\">\n  <span style=\"{T_FOOTER_STRONG}\">📨 {}</span><br />\n  <span>Accept the invitation in your mail client or directly in Nextcloud — your calendar updates either way.</span>\n</div>",
        esc(message)
    )
}

// ── Public renderers ────────────────────────────────────────────

/// Render a Talk-meeting invitation card — the counterpart of
/// `talkInviteHtml` in `inviteHtml.ts`.  Not used by the
/// `create_meeting_invite` composite (its card is the meeting one,
/// with the Talk link as a row), but ported alongside it so the
/// pair of builders stays complete for future tool surfaces.
pub fn talk_invite_html(name: &str, url: &str) -> String {
    let inner = format!(
        "<h1 style=\"{T_TITLE}\">You're invited to a Talk meeting</h1>\n<p style=\"{T_SUBTITLE}\">Click the button below to join the conversation in Nextcloud Talk — works in any modern browser, no install.</p>\n\n<hr style=\"{T_DIVIDER}\" />\n\n{}\n{}\n\n<div style=\"{T_CTA_ROW}\">\n  <a href=\"{}\" style=\"{T_CTA_BUTTON}\">Join Talk meeting →</a>\n</div>\n\n{}",
        detail_row(
            "💬",
            "Talk room",
            &format!("<strong style=\"font-weight:600;\">{}</strong>", esc(name)),
        ),
        detail_row(
            "🔗",
            "Join link",
            &format!(
                "<a href=\"{}\" style=\"color:#3b82f6;text-decoration:none;word-break:break-all;\">{}</a>",
                esc(url),
                esc(url)
            ),
        ),
        esc(url),
        nextcloud_footer("You'll also be added as a participant in Nextcloud Talk."),
    );
    chrome(&inner, "talk-invite")
}

/// Input for [`meeting_invite_html`] — mirrors the TS
/// `MeetingInvite` interface (timestamps arrive as UTC and are
/// rendered in the machine's local timezone, like the webview's
/// `Date` does).
pub(crate) struct MeetingInviteCard {
    pub summary: String,
    pub start: DateTime<Utc>,
    pub end: DateTime<Utc>,
    pub location: Option<String>,
    pub description: Option<String>,
    pub talk_url: Option<String>,
}

/// Render a calendar-meeting invitation card — the counterpart of
/// `meetingInviteHtml` in `inviteHtml.ts`.
pub(crate) fn meeting_invite_html(card: &MeetingInviteCard) -> String {
    meeting_invite_html_in(card, &Local)
}

/// Timezone-parameterised body of [`meeting_invite_html`] so tests
/// can pin a zone and lock exact output bytes.
fn meeting_invite_html_in<Tz: TimeZone>(card: &MeetingInviteCard, tz: &Tz) -> String
where
    Tz::Offset: std::fmt::Display,
{
    let start = card.start.with_timezone(tz);
    let end = card.end.with_timezone(tz);

    let mut detail_rows: Vec<String> = vec![detail_row(
        "📅",
        "When",
        &format!(
            "<strong style=\"font-weight:600;\">{}</strong>",
            esc(&format_range(&start, &end))
        ),
    )];
    if let Some(location) = card.location.as_deref().map(str::trim)
        && !location.is_empty()
    {
        detail_rows.push(detail_row("📍", "Where", &esc(location)));
    }
    let talk_url = card
        .talk_url
        .as_deref()
        .map(str::trim)
        .filter(|u| !u.is_empty());
    if let Some(url) = talk_url {
        detail_rows.push(detail_row(
            "💬",
            "Talk room",
            &format!(
                "<a href=\"{}\" style=\"color:#3b82f6;text-decoration:none;word-break:break-all;\">{}</a>",
                esc(url),
                esc(url)
            ),
        ));
    }
    if let Some(description) = card.description.as_deref().map(str::trim)
        && !description.is_empty()
    {
        // Preserve plain-text newlines as <br /> so an agenda with
        // line breaks reads correctly; escape first so the user's
        // text can't break the surrounding HTML.
        let notes_html = esc(description)
            .replace("\r\n", "<br />")
            .replace('\n', "<br />");
        detail_rows.push(detail_row("📝", "Notes", &notes_html));
    }

    let cta_block = match talk_url {
        Some(url) => format!(
            "<div style=\"{T_CTA_ROW}\">\n  <a href=\"{}\" style=\"{T_CTA_BUTTON}\">Join Talk meeting →</a>\n</div>",
            esc(url)
        ),
        None => String::new(),
    };

    let inner = format!(
        "<h1 style=\"{T_TITLE}\">{}</h1>\n<p style=\"{T_SUBTITLE}\">A calendar invitation has been created in Nextcloud and shared with everyone on this thread.</p>\n\n<hr style=\"{T_DIVIDER}\" />\n\n{}\n\n{}\n\n{}",
        esc(&card.summary),
        detail_rows.join("\n"),
        cta_block,
        nextcloud_footer(
            "You'll be invited via Nextcloud — accepting in your mail client adds the event to your calendar."
        ),
    );
    chrome(&inner, "meeting-invite")
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;

    #[test]
    fn escape_covers_the_five_specials() {
        assert_eq!(
            esc(r#"<a & "b" 'c'>"#),
            "&lt;a &amp; &quot;b&quot; &#39;c&#39;&gt;"
        );
    }

    #[test]
    fn format_range_same_day_and_cross_day() {
        let start = Utc.with_ymd_and_hms(2026, 5, 5, 14, 0, 0).unwrap();
        let end = Utc.with_ymd_and_hms(2026, 5, 5, 15, 0, 0).unwrap();
        assert_eq!(format_range(&start, &end), "Tuesday, May 5 · 14:00 – 15:00");
        let end = Utc.with_ymd_and_hms(2026, 5, 6, 9, 30, 0).unwrap();
        assert_eq!(
            format_range(&start, &end),
            "Tuesday, May 5 14:00 – Wednesday, May 6 09:30"
        );
    }

    #[test]
    fn talk_card_shape_matches_the_ts_builder() {
        let html = talk_invite_html("Weekly sync", "https://cloud.example.com/call/abc123");
        // Wrapper marker + chrome shape the editor's UnkaiBlock
        // extension and the mail clients rely on.
        assert!(html.starts_with("<div data-unkai-block=\"talk-invite\" style=\"max-width:560px;"));
        assert!(html.contains(">Unkai Mail</span>"));
        assert!(html.contains("You're invited to a Talk meeting"));
        assert!(html.contains("Weekly sync"));
        assert!(html.contains("href=\"https://cloud.example.com/call/abc123\""));
        assert!(html.contains("Join Talk meeting →"));
        // The footer message runs through esc() — same as the TS
        // builder — so its apostrophe arrives entity-encoded.
        assert!(html.contains("You&#39;ll also be added as a participant in Nextcloud Talk."));
        // Inline styles only — no <style> block, no <img>, no class=.
        assert!(!html.contains("<style"));
        assert!(!html.contains("<img"));
        assert!(!html.contains("class="));
    }

    #[test]
    fn meeting_card_renders_optional_rows_and_escapes() {
        let card = MeetingInviteCard {
            summary: "Q3 <Planning> & Review".into(),
            start: Utc.with_ymd_and_hms(2026, 5, 5, 14, 0, 0).unwrap(),
            end: Utc.with_ymd_and_hms(2026, 5, 5, 15, 0, 0).unwrap(),
            location: Some("Room 5 & annex".into()),
            description: Some("Agenda:\n1. Numbers\n2. <Plans>".into()),
            talk_url: Some("https://cloud.example.com/call/xyz".into()),
        };
        let html = meeting_invite_html_in(&card, &Utc);
        assert!(
            html.starts_with("<div data-unkai-block=\"meeting-invite\" style=\"max-width:560px;")
        );
        assert!(html.contains("Q3 &lt;Planning&gt; &amp; Review"));
        assert!(html.contains("Tuesday, May 5 · 14:00 – 15:00"));
        assert!(html.contains("Room 5 &amp; annex"));
        assert!(html.contains("Agenda:<br />1. Numbers<br />2. &lt;Plans&gt;"));
        assert!(html.contains("href=\"https://cloud.example.com/call/xyz\""));
        assert!(html.contains("Join Talk meeting →"));
    }

    #[test]
    fn meeting_card_without_talk_url_has_no_cta() {
        let card = MeetingInviteCard {
            summary: "Coffee".into(),
            start: Utc.with_ymd_and_hms(2026, 5, 5, 14, 0, 0).unwrap(),
            end: Utc.with_ymd_and_hms(2026, 5, 5, 15, 0, 0).unwrap(),
            location: None,
            description: None,
            talk_url: None,
        };
        let html = meeting_invite_html_in(&card, &Utc);
        assert!(!html.contains("Join Talk meeting"));
        assert!(!html.contains("📍"));
        assert!(!html.contains("📝"));
        assert!(html.contains("📅"));
    }
}
